Draft: Provide initialization documentation for cloning Dandi ecosystem (#104)
dandi/handbook: docs/62_dandi_cli.md

> For data management (predominately `upload`, `download` and `validation` of data 

Author: @jwodder @yarikoptic would be great if you could review

For reference to easily read in staging setting: https://aquamarine-profiterole-e20e84.netlify.app/62_dandi_cli/

Reviewer: Thanks, Aaron.  Just noting that given your recent developments to push `lincbrain-cli` changes upstream to the `dandi-cli` in https://github.com/dandi/dandi-cli/pull/1519, we will need to update these instructions.

> For data management (predominately `upload`, `download` and `validation` of data 
> to/from DANDI), a local CLI (command line interface) is used.

Reviewer: ```suggestion
Data management including upload, download, and validation is handled through a local DANDI Client.  The DANDI Client provides both a command line interface (CLI) and Python API.
```

Author: Updated!

> 

Reviewer: ```suggestion

For your DANDI Archive clone, you do not need to create a clone of the DANDI Client.  By following the instructions below, your users will be able to utilize the DANDI Client to interface with your DANDI Archive instance.

```

Author: Updated!

> ## Referencing your API

Reviewer: ```suggestion
## Update the DANDI Client to reference your API
```

Author: Updated!

To reference your DANDI-clone API, [update the URLs reference per each CLI action](https://github.com/dandi/dandi-cli/blob/15196a93310618f8897c7b43444e216bbb094549/dandi/consts.py#L119-L135) and push a PR to the [dandi-cli GitHub repository](https://github.com/dandi/dandi-cli).

Here is an example PR of another clone adding to the available instances of `DandiInstance` objects in `dandi-cli`: [see here](https://github.com/dandi/dandi-cli/pull/1527)

For example:

```python
known_instances = {
    "dandi": DandiInstance(
        "dandi",
        "https://dandiarchive.org",
        "https://api.dandiarchive.org/api",
    ),
    "dandi-staging": DandiInstance(
        "dandi-staging",
        "https://gui-staging.<your-domain>.org",
        "https://api-staging.<your-domain>.org/api",
    ),
    "dandi-api-local-docker-tests": DandiInstance(
        "dandi-api-local-docker-tests",
        f"http://{instancehost}:8085",
        f"http://{instancehost}:8000/api",
    ),

>     "<your-dandi-clone>": DandiInstance( # Your own "dandi""

Reviewer: ```suggestion
    "<your-dandi-clone>": DandiInstance( # Your own "dandi"
```

Author: Updated!

        "<your-dandi-clone>", # Your own "dandi"
        "https://<your-dandi-clone-domain>.org", 
        "https://api.<your-dandi-clone-domain>.org/api", 
    ),
}
```

> Once your DANDI clone is added to list of available `DandiInstance` objects, you should be able to perform operations via the `-i` flag -- for example:

Reviewer: ```suggestion
Once your DANDI clone is added to the list of available `DandiInstance` objects, you should be able to perform operations via the `-i` flag -- for example:
```

Author: Updated!

`dandi upload -i <your-dandi-clone>`

> **Note**: Users will be prompted for a `DANDI_API_KEY`
> env. var [see here for code reference](https://github.com/dandi/dandi-cli/blob/6aa414c4db47394970f586cc4fb9758a634aef87/dandi/dandiapi.py#L492-L499)
> 
> You do not need to create a unique value here -- a user can just reference the `DANDI_API_KEY` their API issues in their respective clone.

Reviewer: ```suggestion
## Access credentials

Users will be prompted for a `DANDI_API_KEY`
environment variable.  This variable does not need to be unique to your DANDI clone.  A user can just set their `DANDI_API_KEY` to the value that your DANDI API clone issues.  See docs on [storing access credentials](https://www.dandiarchive.org/handbook/13_upload/#storing-access-credentials).
```

Author: Updated!

> ## Handling Versioning
> 
> DANDI CLI leverages a tool called [versioneer](https://pypi.org/project/versioneer/) for semantic versioning in PyPI.
> 
> Upon merging of a PR into `main`, if a given GitHub label is attached to the PR (`major`, `minor` or `patch` specifically)
> `versioneer` will generate a human-readable CHANGELOG entry, and then push to PyPI the proper new semantic version.
> 
> 
> 
> 
> 

Reviewer: I would suggest removing the documentation on versioneer and just linking to the developer instructions, which includes these details.

```suggestion
## Resources

- [PyPI package](https://pypi.org/project/dandi/)
- [Source code](https://github.com/dandi/dandi-cli)
- [Developer instructions](https://github.com/dandi/dandi-cli/blob/master/DEVELOPMENT.md)
```
```

> 
> 
> 
> 
> 

Reviewer: ```suggestion
```

Data management including upload, download, and validation is handled through a local DANDI Client.  
The DANDI Client provides both a command line interface (CLI) and Python API.

## Update the DANDI Client to reference your API

To reference your DANDI-clone API, [update the URLs reference per each CLI action](https://github.com/dandi/dandi-cli/blob/15196a93310618f8897c7b43444e216bbb094549/dandi/consts.py#L119-L135) and push a PR to the [dandi-cli GitHub repository](https://github.com/dandi/dandi-cli).

> Here is an example PR of another clone adding to the available instances of `DandiInstance` objects in `dandi-cli`: [see here](https://github.com/dandi/dandi-cli/pull/1527)

Reviewer: ```suggestion
Here is an [example PR](https://github.com/dandi/dandi-cli/pull/1527) of another clone adding to the available instances of `DandiInstance` objects in `dandi-cli`.
```

        "https://api.dandiarchive.org/api",
    ),
    "dandi-staging": DandiInstance(
        "dandi-staging",
        "https://gui-staging.<your-domain>.org",
        "https://api-staging.<your-domain>.org/api",
    ),
    "dandi-api-local-docker-tests": DandiInstance(
        "dandi-api-local-docker-tests",
        f"http://{instancehost}:8085",
        f"http://{instancehost}:8000/api",
    ),
    "<your-dandi-clone>": DandiInstance( # Your own "dandi"
        "<your-dandi-clone>", # Your own "dandi"
        "https://<your-dandi-clone-domain>.org", 
        "https://api.<your-dandi-clone-domain>.org/api", 
    ),
}
```

Once your DANDI clone is added to the list of available `DandiInstance` objects, you should be able to perform operations via the `-i` flag.  For example:

`dandi upload -i <your-dandi-clone>`

## Access credentials

Users will be prompted for a `DANDI_API_KEY`
environment variable.  This variable does not need to be unique to your DANDI clone.  A user can just set their `DANDI_API_KEY` to the value that your DANDI API clone issues.  See docs on [storing access credentials](https://www.dandiarchive.org/handbook/13_upload/#storing-access-credentials).

> ## Handling Versioning

Reviewer: ```suggestion
## Versioning
```

> DANDI CLI leverages a tool called [versioneer](https://pypi.org/project/versioneer/) for semantic versioning in PyPI.

Reviewer: ```suggestion
The DANDI Client leverages a tool called [versioneer](https://pypi.org/project/versioneer/) for semantic versioning in PyPI.
```

> Upon merging of a PR into `main`, if a given GitHub label is attached to the PR (`major`, `minor` or `patch` specifically)
> `versioneer` will generate a human-readable CHANGELOG entry, and then push to PyPI the proper new semantic version.

Reviewer: ```suggestion
Upon merging of a PR into `master`, if the `release` label is attached to the PR
`versioneer` will generate a human-readable CHANGELOG entry, and then push to PyPI the proper new semantic version.  For more details on labeling `dandi-cli` pull requests, see [here](https://github.com/dandi/dandi-cli/blob/master/DEVELOPMENT.md#releasing-with-github-actions-auto-and-pull-requests).
```

> ### Versioning-Realted Resources

Reviewer: ```suggestion
### Resources
```
